src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,107 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIAL_CAPACITY = 10;
+    private int size;
+    private Object[] elementData;
+
+    public ArrayList() {
+        elementData = new Object[INITIAL_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            growElementData();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (size == index) {
+            add(value);
+            return;
+        }
+        if (index > -1 && index < size) {
+            if (size + 1 == elementData.length) {
+                growElementData();
+            }
+            System.arraycopy(elementData, index, elementData, index + 1, size - index);
+            elementData[index] = value;
+            size++;
+            return;
+        }
+        throw new ArrayListIndexOutOfBoundsException("Wrong index!");
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int totalLengthOfMergedArrays = elementData.length + list.size();
+        while (elementData.length < totalLengthOfMergedArrays) {
+            growElementData();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index == size) {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index!");
+        }
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index == size) {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index!");
+        }
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T elementToReturn = get(index);
+        if (index == size - 1) {
+            elementData[index] = null;
+            size--;
+            return elementToReturn;
+        }
+        System.arraycopy(elementData, index + 1, elementData, index, size - index);
+        size--;
+        return elementToReturn;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if ((elementData[i] == null && element == null)
+                    || (elementData[i] != null && elementData[i].equals(element))) {
```

Comment:
unnecessary parentheses 
can be simplified
```suggestion
            if (elementData[i] == element
                    || elementData[i] != null && elementData[i].equals(element)) {
```

Reply:
Done